synapse/rest/client/v2_alpha/account.py

```diff
@@ -512,35 +683,35 @@ def on_GET(self, request):
 
     @defer.inlineCallbacks
     def on_POST(self, request):
+        requester = yield self.auth.get_user_by_req(request)
+        user_id = requester.user.to_string()
         body = parse_json_object_from_request(request)
 
         threepid_creds = body.get("threePidCreds") or body.get("three_pid_creds")
         if threepid_creds is None:
             raise SynapseError(
                 400, "Missing param three_pid_creds", Codes.MISSING_PARAM
             )
+        assert_params_in_dict(threepid_creds, ["client_secret", "sid"])
 
-        requester = yield self.auth.get_user_by_req(request)
-        user_id = requester.user.to_string()
-
-        # Specify None as the identity server to retrieve it from the request body instead
-        threepid = yield self.identity_handler.threepid_from_creds(None, threepid_creds)
+        client_secret = threepid_creds["client_secret"]
+        sid = threepid_creds["sid"]
 
-        if not threepid:
-            raise SynapseError(400, "Failed to auth 3pid", Codes.THREEPID_AUTH_FAILED)
+        # Get a validated session matching these details
+        validation_session = yield self.datastore.get_threepid_validation_session(
+            None, client_secret, sid=sid, validated=True
+        )
 
-        for reqd in ["medium", "address", "validated_at"]:
-            if reqd not in threepid:
-                logger.warn("Couldn't add 3pid: invalid response from ID server")
-                raise SynapseError(500, "Invalid response from ID Server")
+        if not validation_session:
+            raise SynapseError(
+                400, "No validated 3pid session found", Codes.THREEPID_AUTH_FAILED
+            )
 
-        yield self.auth_handler.add_threepid(
-            user_id, threepid["medium"], threepid["address"], threepid["validated_at"]
-        )
+        address = validation_session["address"]
+        medium = validation_session["medium"]
+        validated_at = validation_session["validated_at"]
 
-        if "bind" in body and body["bind"]:
```

Comment:
it feels to me like doing the validation emails on the HS, and dropping support for `bind`, are orthogonal changes? do they need to happen at the same time?

Reply:
more specifically: we could drop support for `bind` first, and *then* implement sending our own email.

Reply:
Sure, I can split out the `bind` dropping into another PR.

Reply:
https://github.com/matrix-org/synapse/pull/6067